crates/pathfinder/src/sequencer/error.rs

```diff
@@ -0,0 +1,103 @@
+pub enum SequencerError {
+    /// All errors related to parsing sequencer replies that should
+    /// be in the JSON format.
+    #[error("Failed to parse sequencer reply JSON: {0}")]
+    DeserializationError(#[from] serde_json::Error),
+    /// All errors related to parsing sequencer replies that
+    /// are not related to JSON handling.
+    #[error("Failed to parse a non-JSON sequencer reply: {0}")]
+    ParseError(#[from] anyhow::Error),
+    /// Starknet specific errors.
+    #[error("Starknet error: {0}")]
+    StarknetError(#[from] StarknetError),
+    /// Networking and protocol related errors.
+    #[error("Sequencer transport error: {0}")]
+    TransportError(#[from] reqwest::Error),
+}
+
+impl From<SequencerError> for rpc::Error {
+    fn from(e: SequencerError) -> Self {
+        match e {
+            SequencerError::DeserializationError(e) => {
+                rpc::Error::Call(rpc::CallError::Failed(e.into()))
+            }
+            SequencerError::ParseError(e) => rpc::Error::Call(rpc::CallError::Failed(e)),
+            SequencerError::StarknetError(e) => match e.code {
+                StarknetErrorCode::OutOfRangeBlockHash | StarknetErrorCode::BlockNotFound => {
+                    RpcErrorCode::InvalidBlockHash.into()
+                }
+                StarknetErrorCode::OutOfRangeContractAddress
+                | StarknetErrorCode::UninitializedContract => RpcErrorCode::ContractNotFound.into(),
+                StarknetErrorCode::OutOfRangeTransactionHash => {
+                    rpc::Error::Call(rpc::CallError::Custom {
+                        code: RpcErrorCode::InvalidTransactionHash as i32,
+                        message: RpcErrorCode::InvalidTransactionHashStr.to_owned(),
+                        data: None,
+                    })
+                }
+                StarknetErrorCode::OutOfRangeStorageKey => RpcErrorCode::InvalidStorageKey.into(),
+                StarknetErrorCode::TransactionFailed => RpcErrorCode::InvalidCallData.into(),
+                StarknetErrorCode::EntryPointNotFound => {
+                    RpcErrorCode::InvalidMessageSelector.into()
+                }
+                StarknetErrorCode::MalformedRequest
+                    if e.message.contains("Block ID should be in the range") =>
+                {
+                    RpcErrorCode::InvalidBlockNumber.into()
+                }
+                _ => rpc::Error::Call(rpc::CallError::Failed(e.into())),
+            },
+            SequencerError::TransportError(e) => rpc::Error::Call(rpc::CallError::Failed(e.into())),
+        }
+    }
+}
+
+/// Used for deserializing specific Starknet sequencer error data.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct StarknetError {
+    pub code: StarknetErrorCode,
+    pub message: String,
+    // The `problems` field is intentionally omitted here
+    // Let's deserialize it if it proves necessary
+}
+
+impl std::error::Error for StarknetError {}
```

Comment:
```suggestion
impl Error for StarknetError {}
```

Reply:
I think `Error` is in the `std::prelude::*`

Reply:
Maybe I used the prefix to avoid ambiguity - I'll check it.

Reply:
Hmm, ok, now I know why:
```
error[E0405]: cannot find trait `Error` in this scope
  --> crates/pathfinder/src/sequencer/error.rs:71:6
   |
71 | impl Error for StarknetError {}
   |      ^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
2  | use serde::de::Error;
   |
2  | use serde::ser::Error;
   |
2  | use std::error::Error;

```

Reply:
Hmm okay nevermind then 👍